lib/scholar/cluster/affinity_propagation.ex

```diff
@@ -0,0 +1,321 @@
+
+        y2 = Nx.reduce_max(temp, axes: [1])
+
+        temp = s - Nx.new_axis(y, -1)
+
+        temp =
+          Nx.indexed_put(
+            temp,
+            max_indices,
+            Nx.flatten(Nx.gather(s, max_indices) - y2)
+          )
+
+        temp = temp * (1 - damping_factor)
+
+        r = r * damping_factor + temp
+
+        temp = Nx.max(r, 0)
+
+        temp = Nx.put_diagonal(temp, Nx.take_diagonal(r))
+
+        temp = temp - Nx.sum(temp, axes: [0])
+
+        a_change = Nx.take_diagonal(temp)
+
+        temp = Nx.clip(temp, 0, inf)
+
+        temp = Nx.put_diagonal(temp, a_change)
+
+        temp = temp * (1 - damping_factor)
+
+        a = a * damping_factor - temp
+
+        diagonals = Nx.take_diagonal(a) + Nx.take_diagonal(r) > 0
+
+        {a, r, diagonals, s, i + 1}
+      end
+
+    {s, affinity_matrix, diagonals, data}
+  end
+
+  @doc """
+  Computes labels and centers of clusterization.
+
+  ## Return Values
+
+    The function returns a struct with the following parameters:
+
+    * `:affinity_matrix` - Affinity matrix. It is a negated squared euclidean distance of each pair of points.
+
+    * `:clusters_centers` - Cluster centers from the initial data.
+
+    * `:cluster_centers_indices` - Indices of cluster centers.
+
+    * `:labels` - Labels of each point.
+
+  ## Examples
+
+      iex> seed = 42
+      iex> x = Nx.tensor([[12,5,78,2], [1,5,7,32], [1,3,6,1], [1,2,5,2]])
+      iex> {s, affinity_matrix, diagonals, data} = Scholar.Cluster.AffinityPropagation.fit(x, seed: seed)
+      iex> Scholar.Cluster.AffinityPropagation.compute_values({s, affinity_matrix, diagonals, data})
+      %Scholar.Cluster.AffinityPropagation{
+        labels: Nx.tensor([0, 1, 1, 1]),
+        cluster_centers_indices: Nx.tensor([0, 3]),
+        affinity_matrix: Nx.tensor(
+          [
+            [0.0, -6062.0, -5310.0, -5459.0],
+            [-6062.0, 0.0, -966.0, -913.0],
+            [-5310.0, -966.0, 0.0, -3.0],
+            [-5459.0, -913.0, -3.0, 0.0]
+          ]),
+        cluster_centers: Nx.tensor(
+          [
+            [12, 5, 78, 2],
+            [1, 2, 5, 2]
+          ]
+        )
+      }
+  """
+  def compute_values({s, affinity_matrix, diagonals, data}) do
+    k = Nx.sum(diagonals, axes: [0]) |> Nx.to_number()
+
+    {cluster_centers, cluster_centers_indices, labels} =
+      if k > 0 do
+        indices =
+          Nx.multiply(Nx.flatten(diagonals), Nx.add(Nx.iota(Nx.shape(diagonals)), 1))
+          |> Nx.to_flat_list()
+          |> Enum.filter(fn x -> x != 0 end)
+          |> Nx.tensor()
+          |> Nx.subtract(1)
+          |> Nx.as_type({:s, 64})
+
+        c = Nx.argmax(Nx.take(s, indices, axis: 1), axis: 1)
+        c = Nx.indexed_put(c, Nx.new_axis(indices, -1), Nx.iota({k}))
+        {Nx.take(data, indices), indices, c}
+      else
+        {n, _} = Nx.shape(s)
+
+        {Nx.tensor([-1], type: Nx.type(s)), Nx.broadcast(Nx.tensor(-1, type: Nx.type(s)), {n}),
```

Comment:
Is `indices` always meant to have size `{n}`? Or it may change?

Reply:
No, I'll introduce some negative values.

Reply:
No, it may change. The case in the else part indicates that the algorithm does not converge to any solution and we need to return some "trash".

Reply:
I see. I have a question: could we then implement this whole function as a `defn`? Then in the `if true` case, we may have some `-1` values in the indexes. Then an operation, outside of `defn`, could do the `take` only on the indices that are not negative?

Reply:
I abstract the function compute_values from fit because I couldn't implement this part in `defn`, otherwise there wouldn't be that function. The problem with defn in this situation is that algorithm decides on its own how many clusters it predict, and that introduces dynamic shapes (that is why I use Enum.filter).

Reply:
Yes, what I propose is for us to compute and return fixed size tensors in `fit`, using only `defn`, and use `-1` indexes to say which entries are dummies. Something like this:

   %Scholar.Cluster.AffinityPropagation{
        labels: Nx.tensor([0, 1, 1, 1]),
        cluster_centers_indices: Nx.tensor([0, -1, -1, 3]),
        affinity_matrix: Nx.tensor(
          [
            [0.0, -6162.0, -5358.0, -5499.0],
            [-6162.0, 0.0, -1030.0, -913.0],
            [-5358.0, -1030.0, 0.0, -31.0],
            [-5499.0, -913.0, -31.0, 0.0]
          ]),
        cluster_centers: Nx.tensor(
          [
            [12, 5, 78, 2],
            [0, 0, 0, 0],
            [0, 0, 0, 0],
            [1, -2, 5, 2]
          ]
        )
      }

We could even have a field called `num_clusters` which returns the number of actual clusters. Then we can have a function called `prune`, which is not a `defn`, that receives the struct above and remove all negative indexes and the dummy clusters.

Reply:
Ok, I can recreate `cluster_centers_indices` and `cluster_centers` in the proposed manner, but labels cannot be calculated inside defn.

Reply:
Where would we use the labels? They are not used in `predict` right now, are they?

Reply:
Well, labels are the 'final product' and they are not used in `predict`. I thought I implemented an alternative version with static shapes to compute labels, but some labels are wrong (And tbh I don't know why). 

working version, but with dynamic shapes
```elixir
labels = Nx.argmax(Nx.take(s, cluster_centers_indices, axis: 1), axis: 1)
labels = Nx.indexed_put(labels, Nx.new_axis(cluster_centers_indices, -1), Nx.iota({k}))
```

my alternative version
```elixir
  c =
    Nx.select(
      Nx.broadcast(Nx.new_axis(Nx.flatten(diagonals) != 0, -1), Nx.shape(s)),
      s,
      Nx.Constants.neg_infinity(Nx.type(s))
    )

  c = Nx.argmax(c, axis: 0) |> Nx.as_type({:s, 64})
```

Reply:
What if make this in steps? What if we do this `fit |> compute_values |> predict`, but `fit` already computes the struct and `compute_values` is a regular function that simply removes the negative indexes?

Reply:
It is a good first step. We can look into improving it further next :)

Reply:
Btw @josevalim, are the warnings that I've got after the most recent pushes correct? Because I don't think so, if I change -1 to 1 in tensors then warnings magically disappear.

Reply:
Inside defn the unary - operator becomes the negate function. I think we could get around this by using Kernel.- if the input is a number